core/contextmenu_registry.ts

```diff
@@ -125,33 +114,71 @@ export class ContextMenuRegistry {
     return menuOptions;
   }
 }
-export interface Scope {
-  block: BlockSvg|undefined;
-  workspace: WorkspaceSvg|undefined;
-}
-export interface RegistryItem {
-  callback: (p1: Scope) => AnyDuringMigration;
-  scopeType: ScopeType;
-  displayText: ((p1: Scope) => string)|string;
-  preconditionFn: (p1: Scope) => string;
-  weight: number;
-  id: string;
-}
-export interface ContextMenuOption {
-  text: string;
-  enabled: boolean;
-  callback: (p1: Scope) => AnyDuringMigration;
-  scope: Scope;
-  weight: number;
-}
-export interface LegacyContextMenuOption {
-  text: string;
-  enabled: boolean;
-  callback: (p1: Scope) => AnyDuringMigration;
+
+export namespace ContextMenuRegistry {
+  /**
+   * Where this menu item should be rendered. If the menu item should be
+   * rendered in multiple scopes, e.g. on both a block and a workspace, it
+   * should be registered for each scope.
+   */
+  export enum ScopeType {
+    BLOCK = 'block',
+    WORKSPACE = 'workspace',
+  }
+
+  /**
+   * The actual workspace/block where the menu is being rendered. This is passed
+   * to callback and displayText functions that depend on this information.
+   */
+  export interface Scope {
+    block: BlockSvg|undefined;
+    workspace: WorkspaceSvg|undefined;
+  }
+
+  /**
+   * A menu item as entered in the registry.
+   */
+  export interface RegistryItem {
+    callback: (p1: Scope) => AnyDuringMigration;
+    scopeType: ScopeType;
+    displayText: ((p1: Scope) => string)|string;
+    preconditionFn: (p1: Scope) => string;
+    weight: number;
+    id: string;
+  }
+
+  /**
+   * A menu item as presented to contextmenu.js.
+   */
+  export interface ContextMenuOption {
+    text: string;
+    enabled: boolean;
+    callback: (p1: Scope) => AnyDuringMigration;
+    scope: Scope;
+    weight: number;
+  }
+
+  /**
+   * A subset of ContextMenuOption corresponding to what was publicly
+   * documented.  ContextMenuOption should be preferred for new code.
+   */
+  export interface LegacyContextMenuOption {
+    text: string;
+    enabled: boolean;
+    callback: (p1: Scope) => AnyDuringMigration;
+  }
+
+  /**
+   * Singleton instance of this class. All interactions with this class should
+   * be done on this object.
+   */
+  ContextMenuRegistry.registry = new ContextMenuRegistry();
 }
 
-/**
- * Singleton instance of this class. All interactions with this class should be
- * done on this object.
- */
-ContextMenuRegistry.registry = new ContextMenuRegistry();
+export type ScopeType = ContextMenuRegistry.ScopeType;
```

Comment:
These are only available internally to Blockly core correct? We're not planning to export these from the top level?

Reply:
Not at the present time, no.

We could export them under their own names later if we want to.

Or alternatively we could remove these separate exports and change the code that MigranTS changed back to how it used to be (using the `static`-property versions of the types).